### PR TITLE
TypeScript project references and Nx Cloud

### PR DESCRIPTION
**Pull request by adriencaccia**
- Implement a correct usage of TypeScript project references using this article: https://moonrepo.dev/docs/guides/javascript/typescript-project-refs
- Setup Nx Cloud

### CODE REVIEW

**Comment by adriencaccia** [comment]
@fargito WDYT?

File: nx.json [hunk 1]
```diff
@@ -14,7 +14,8 @@
           "test-circular",
           "test-unit"
         ],
-        "cacheDirectory": "nx-cache"
+        "cacheDirectory": "nx-cache",
+        "accessToken": "MzY4MGJjNWEtZmE5YS00ZWQ2LTgzOTEtYzY0Y2MzOTFlMGJkfHJlYWQtd3JpdGU="
```
Comment (fargito) [inline nx.json:18]: @adriencaccia is this a real one?

File: tsconfig.json [hunk 2]
```diff
@@ -1,22 +1,18 @@
...
+    { "path": "./packages/create-swarmion-app/tsconfig.json" },
+    { "path": "./packages/eslint-plugin/tsconfig.build.json" },
+    { "path": "./packages/nx-plugin/tsconfig.build.json" },
+    { "path": "./packages/serverless-cdk-plugin/tsconfig.build.json" },
+    { "path": "./packages/serverless-configuration/tsconfig.build.json" },
+    { "path": "./packages/serverless-contracts/tsconfig.build.json" },
+    { "path": "./packages/serverless-contracts-plugin/tsconfig.build.json" },
+    { "path": "./packages/serverless-helpers/tsconfig.build.json" },
+    { "path": "./user-docs/cloudfront/tsconfig.json" },
+    { "path": "./user-docs/documentation/tsconfig.json" }
```
Comment (fargito) [inline tsconfig.json:16]: This means we have to update this file when generating a new package/service right?

File: contracts/orchestrator-contracts/tsconfig.json [hunk 3]
```diff
@@ -1,10 +1,13 @@
 {
-  "extends": "../../tsconfig.json",
+  "extends": "../../tsconfig.options.json",
```
Comment (fargito) [inline contracts/orchestrator-contracts/tsconfig.json:2]: Shouldn't this change be in `tsconfig.build.json` instead?

**Review by fargito — CHANGES_REQUESTED** [review]
That would be cleaner indeed!

However, I'm confused to whether this change must apply only for `tsconfig.json`, `tsconfig.build.json` or both in libraries.

Also, we need to update generators as well to reflect this change.

File: .gitignore [hunk 4]
```diff
@@ -43,3 +43,6 @@ event.json
 
 # Nx caching
 nx-cache
+
+# Build cache manifests
+*.tsbuildinfo
```
Comment (fargito) [inline .gitignore:48]: I'm also wondering whether this should be cached by Nx as well. In this case, it should be included in the root Nx config

File: nx.json (hunk 1 above)
Comment (adriencaccia) [reply nx.json:18]: It is yes!

File: tsconfig.json (hunk 2 above)
Comment (adriencaccia) [reply tsconfig.json:16]: Yes, that way we can run [`tsc --build --verbose` at the root-level](https://moonrepo.dev/docs/guides/javascript/typescript-project-refs#on-all-projects)

File: .gitignore (hunk 4 above)
Comment (adriencaccia) [reply .gitignore:48]: You are completely right 😉

File: contracts/orchestrator-contracts/tsconfig.json (hunk 3 above)
Comment (adriencaccia) [reply contracts/orchestrator-contracts/tsconfig.json:2]: I followed [the guide](https://moonrepo.dev/docs/guides/javascript/typescript-project-refs#on-all-projects), and everything is working as expected

**Comment by adriencaccia** [comment]
> That would be cleaner indeed!
> 
> However, I'm confused to whether this change must apply only for `tsconfig.json`, `tsconfig.build.json` or both in libraries.
> 
> Also, we need to update generators as well to reflect this change.

To have a better understanding of the impacts, I would like to test on a bigger repo with known issues (@CorentinDoue I am looking at yours 👀). I will try to migrate it the same way I did in this PR and check if it solves some issues.

File: tsconfig.json (hunk 2 above)
Comment (fargito) [reply tsconfig.json:16]: This needs to be populated in generators right?

File: packages/create-swarmion-app/package.json [hunk 5]
```diff
@@ -35,7 +35,7 @@
     "test": "pnpm test-linter && pnpm test-type && pnpm test-unit && pnpm test-circular",
     "test-circular": "pnpm depcruise --validate .dependency-cruiser.js .",
     "test-linter": "pnpm linter-base-config .",
-    "test-type": "tsc --noEmit --emitDeclarationOnly false",
+    "test-type": "tsc",
```
Comment (guillaumeduboc) [inline packages/create-swarmion-app/package.json:38]: why don't we do this modification in every package ?

File: contracts/orchestrator-contracts/tsconfig.json [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
-    "baseUrl": "src",
-    "composite": true,
-    "outDir": "./dist/types"
+    "baseUrl": "src"
   },
+  "references": [
+    {
+      "path": "../../packages/serverless-contracts/tsconfig.build.json"
+    }
+  ],
```
Comment (guillaumeduboc) [inline contracts/orchestrator-contracts/tsconfig.json:10]: is this still needed ? 
it is already specified in the root `tsconfig.options.json`

File: packages/create-swarmion-app/package.json (hunk 5 above)
Comment (adriencaccia) [reply packages/create-swarmion-app/package.json:38]: To make `tsc --build` work in both packages and apps, we need to differentiate.
In apps, we will have `"test-type": "tsc",` and `"test-type": "tsc --noEmit --emitDeclarationOnly false",` in packages.

File: tsconfig.json (hunk 2 above)
Comment (adriencaccia) [reply tsconfig.json:16]: I upgrade them!

**Comment by adriencaccia** [comment]
@fargito It's done 🎉 
I updated the generators and modified the examples.
I didn't have time to test on @CorentinDoue's project.
But I think we can merge anyway, the cleaning of the configs in itself is a good step forward.
And there are other cool things about PR 😉 
WDYT?

**Review by fargito — COMMENTED** [review]
LGTM, thanks a lot for the hard work! :rocket:

**Review by fargito — APPROVED** [review]

**Pull request merged**